File: backend_memory.go
package bbq

import (
	"errors"
	"math/rand"
	"time"
)

type MemoryBackend struct {
	vecs []Vector
	rng  *rand.Rand
	dim  int
}

var _ scannableBackend = &MemoryBackend{}
var _ VectorGetter[Vector] = &MemoryBackend{}

func NewMemoryBackend(dimensions int) *MemoryBackend {
	return &MemoryBackend{
		rng: rand.New(rand.NewSource(time.Now().UnixMicro())),
		dim: dimensions,
	}
}

func (mem *MemoryBackend) PutVector(id ID, vector Vector) error {
	if len(vector) != mem.dim {
		return errors.New("MemoryBackend: vector dimension doesn't match")
	}

	if int(id) < len(mem.vecs) {
		mem.vecs[int(id)] = vector
	} else if int(id) == len(mem.vecs) {
		mem.vecs = append(mem.vecs, vector)
	} else {
		mem.grow(int(id))
		mem.vecs[int(id)] = vector
	}
	return nil
}

func (mem *MemoryBackend) grow(to int) {
	diff := (to - len(mem.vecs)) + 1
	mem.vecs = append(mem.vecs, make([]Vector, diff)...)
}

func (mem *MemoryBackend) ComputeSimilarity(vector Vector, targetID ID) (float32, error) {
	target, err := mem.GetVector(targetID)
	if err != nil {
		return 0, err
	}
	return target.CosineSimilarity(vector), nil
}

func (mem *MemoryBackend) Info() BackendInfo {
	return BackendInfo{
		HasIndexData: false,
		Dimensions:   mem.dim,
		VectorCount:  len(mem.vecs),
	}
}

func (mem *MemoryBackend) GetVector(id ID) (Vector, error) {
	if int(id) > len(mem.vecs)-1 {
		return nil, ErrIDNotFound
	}
	if mem.vecs[int(id)] == nil {
		return nil, ErrIDNotFound
	}
	return mem.vecs[int(id)], nil
}

func (mem *MemoryBackend) Exists(id ID) bool {
	i := int(id)
	if len(mem.vecs) <= i {
		return false
	}
	return mem.vecs[i] != nil
}

func (mem *MemoryBackend) ForEachVector(cb func(ID) error) error {
	for i, v := range mem.vecs {
		if v == nil {
			continue
		}
		err := cb(ID(i))
		if err != nil {
			return err
		}
	}
	return nil
}

File: rust/src/backend_memory.rs
use std::{
    cmp::min,
    sync::{Arc, Mutex},
};

use anyhow::{anyhow, Result};
use rand::RngCore;

use crate::{
    backend::{BackendInfo, VectorBackend},
    quantization::Quantization,
    Vector, ID,
};

pub struct QuantizedMemoryBackend<Q: Quantization> {
    vecs: Vec<Option<Q::Lower>>,
    dimensions: usize,
    n_basis: usize,
    rng: Option<Arc<Mutex<Box<dyn RngCore + Send>>>>,
}

pub type MemoryBackend = QuantizedMemoryBackend<crate::quantization::NoQuantization>;

impl<Q: Quantization> QuantizedMemoryBackend<Q> {
    pub fn new(dimensions: usize, n_basis: usize) -> Result<Self> {
        Ok(Self {
            vecs: Vec::new(),
            dimensions,
            n_basis,
            rng: None,
        })
    }

    pub fn set_rng(&mut self, rng: Box<dyn RngCore + Send>) {
        self.rng = Some(Arc::new(Mutex::new(rng)));
    }
}

impl<Q: Quantization> VectorBackend for QuantizedMemoryBackend<Q> {
    fn put_vector(&mut self, id: crate::ID, v: &Vector) -> Result<()> {
        if v.len() != self.dimensions {
            return Err(anyhow!("dimensions don't match"));
        }
        let uid = id as usize;
        if self.vecs.len() <= uid {
            if self.vecs.capacity() == uid {
                self.vecs.reserve(min(self.vecs.capacity(), 1024 * 1024))
            }
            self.vecs.resize(uid + 1, None);
        }
        let mut insert = v.clone();
        crate::vector::normalize(&mut insert);
        let l = Q::lower(insert)?;
        self.vecs[uid] = Some(l);
        Ok(())
    }

    fn compute_similarity(&self, target: &Vector, target_id: crate::ID) -> Result<f32> {
        // Make sure it's normalized!
        let v = self.vecs[target_id as usize]
            .as_ref()
            .ok_or(anyhow!("No vector present"))?;
        Q::compare(target, v)
    }

    fn info(&self) -> crate::backend::BackendInfo {
        BackendInfo {
            has_index_data: false,
            dimensions: self.dimensions,
            n_basis: self.n_basis,
            vector_count: self.vecs.len(),
            quantization: Q::name().into(),
        }
    }

    fn iter_vector_ids(&self) -> impl Iterator<Item = ID> {
        self.vecs
            .iter()
            .enumerate()
            .filter(|(_, v)| v.is_some())
            .map(|(k, _)| k as ID)
    }

    fn vector_exists(&self, id: ID) -> bool {
        let v = self.vecs.get(id as usize);
        match v {
            Some(x) => x.is_some(),
            None => false,
        }
    }

    fn close(self) -> Result<()> {
        todo!()
    }

    fn load_bases(&self) -> Result<Option<Vec<crate::Basis>>> {
        Ok(None)
    }

    fn load_bitmap<T: crate::Bitmap>(&mut self, _basis: usize, _index: i32) -> Result<Option<T>> {
        Ok(None)
    }
}
